src/adapter/src/catalog.rs

```diff
@@ -2711,17 +2711,23 @@ impl Catalog {
             builtin_table_updates.push(catalog.state.pack_audit_log_update(&event)?);
         }
 
-        // To avoid reading over storage_usage() multiple times, do both the
-        // table updates and most-recent-timestamp calculations on a single
-        // iterator.
-        let storage_usage_events = catalog.storage().await.storage_usage().await?;
         // If no usage retention period is set, we set it to an unreasonably large number
         // of milliseconds, so that the filter code is simpler.
         let cutoff_ts = match config.storage_usage_retention_period {
             None => u128::MIN,
             Some(period) => u128::from(catalog.storage().await.boot_ts()) - period.as_millis(),
         };
-        for event in storage_usage_events.filter(|e| u128::from(e.timestamp()) > cutoff_ts) {
+        catalog
+            .storage()
+            .await
+            .prune_storage_usage(move |e| u128::from(e.timestamp()) <= cutoff_ts)
+            .await?;
+
+        // To avoid reading over storage_usage() multiple times, do both the
```

Comment:
Due to the call to `.delete` above, we now ARE reading over the storage usage table multiple times. What I would like to do is: call `catalog.storage().await.storage_usage()` once. Split it into two datasets: 1) things to be deleted, 2) things to be emitted to the builtin tables. Then we call `COLLECTION_STORAGE_USAGE.transaction` (which doesn't read any rows from the stash), and create an append batch that we call with `txn.append`. This is what `TypedCollection.delete` does under the covers, so you could look at that function and mostly just copy it (you can ignore the PeekSinceUpper stuff).

Alternatively: change this comment to be a TODO to make it read only once, and I'll do the above stuff in a followup PR.